File: iosApp/iosApp/Modules/KeyStore/KeyStoreInteractor.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3lib

protocol KeyStoreInteractor: AnyObject {
    var selected: KeyStoreItem? { set get}
    var items: [KeyStoreItem] { get }
    func add(item: KeyStoreItem, password: String, secretStorage: SecretStorage)
}

final class DefaultKeyStoreInteractor {
    private var keyStoreService: KeyStoreService
    private var networksService: NetworksService

    init(
        _ keyStoreService: KeyStoreService,
        networksService: NetworksService
    ) {
        self.keyStoreService = keyStoreService
        self.networksService = networksService
    }
}

extension DefaultKeyStoreInteractor: KeyStoreInteractor {
    
    var selected: KeyStoreItem? {
        get { keyStoreService.selected }
        set { setSelected(newValue) }
    }

    var items: [KeyStoreItem] {
        keyStoreService.items()
    }

    func add(item: KeyStoreItem, password: String, secretStorage: SecretStorage) {
        keyStoreService.add(item: item, password: password, secretStorage: secretStorage)
    }
}

private extension DefaultKeyStoreInteractor {

    func setSelected(_ keyStoreItem: KeyStoreItem?) {
        keyStoreService.selected = keyStoreItem
        guard let keyStoreItem = keyStoreItem else { return }
        networksService.keyStoreItem = keyStoreItem
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/keyStore/KeyStoreInteractor.kt
package com.sonsofcrypto.web3walletcore.modules.keyStore

import com.sonsofcrypto.web3lib.services.keyStore.KeyStoreItem
import com.sonsofcrypto.web3lib.services.keyStore.KeyStoreService
import com.sonsofcrypto.web3lib.services.keyStore.SecretStorage
import com.sonsofcrypto.web3lib.services.networks.NetworksService

interface KeyStoreInteractor {
    var selected: KeyStoreItem?
    val items: List<KeyStoreItem>
    fun add(item: KeyStoreItem, password: String, secretStorage: SecretStorage)
}

class DefaultKeyStoreInteractor(
    private var keyStoreService: KeyStoreService,
    private var networksService: NetworksService,
): KeyStoreInteractor {

    override var selected: KeyStoreItem?
        get() = keyStoreService.selected
        set(value) {
            keyStoreService.selected = value
            if (value != null) {
                networksService.keyStoreItem = value
            }
        }

    override val items: List<KeyStoreItem> get() = keyStoreService.items()

    override fun add(item: KeyStoreItem, password: String, secretStorage: SecretStorage) {
        keyStoreService.add(item, password, secretStorage)
    }
}
